icicle/backend/cpu/include/cpu_program_executor.h

```diff
@@ -0,0 +1,96 @@
+#pragma once
+
+#include <vector>
+#include <functional>
+#include "icicle/program/symbol.h"
+#include "icicle/program/program.h"
+
+namespace icicle {
+
+  template <typename S>
+  class CpuProgramExecutor
+  {
+  public:
+    CpuProgramExecutor(Program<S>& program)
+        : m_program(program), m_variable_ptrs(program.get_nof_vars()), m_intermidites(program.m_nof_intermidiates)
+    {
+      // initialize m_variable_ptrs vector
+      int variable_ptrs_idx = program.m_nof_inputs + program.m_nof_outputs;
+      for (int idx = 0; idx < program.m_nof_constants; ++idx) {
+        m_variable_ptrs[variable_ptrs_idx++] = &(program.m_constants[idx]);
+      }
+      for (int idx = 0; idx < program.m_nof_intermidiates; ++idx) {
+        m_variable_ptrs[variable_ptrs_idx++] = &(m_intermidites[idx]);
+      }
+    }
+
+    // execute the program an return a program to the result
+    void execute()
+    {
+      const std::byte* instruction;
+      for (InstructionType instruction : m_program.m_instructions) {
+        const int func_select = (instruction & 0xFF);
```

Comment:
consider adding macros to extract fields and make the code more readable

Reply:
done